Add #create_or_find_by to lean on unique constraints

dhh opened the pull request:
Attempts to create a record with the given attributes in a table that has a unique constraint
on one or several of its columns. If a row already exists with one or several of these
unique constraints, the exception such an insertion would normally raise is caught, 
and the existing record with those attributes is sought found using #find_by.

This is similar to #find_or_create_by, but avoids the problem of stale reads, as that methods needs
to first query the table, then attempt to insert a row if none is found. That leaves a timing gap
between the SELECT and the INSERT statements that can cause problems in high throughput applications.

There are several drawbacks to #create_or_find_by, though:

* The underlying table must have the relevant columns defined with unique constraints.
* A unique constraint violation may be triggered by only one, or at least less than all, 
  of the given attributes. This means that the subsequent #find_by may fail to find a
  matching record, which will then return nil, rather than a record will the given attributes.
* It relies on exception handling to handle control flow, which may be marginally slower. And

This method will always returns a record if all given attributes are covered by unique constraints,
but if creation was attempted and failed due to validation errors it won't be persisted, you get what
#create returns in such situation.

Commit:
Add #create_or_find_by to lean on unique constraints

nynhex commented:
This is really good.  I have multiple use cases for this.  👍

jeremy commented on activerecord/lib/active_record/relation.rb:
```diff
@@ -170,6 +159,41 @@ def find_or_create_by!(attributes, &block)
+    def create_or_find_by!(attributes, &block)
+      create!(attributes, &block)
+    rescue ActiveRecord::RecordNotUnique
+      find_by(attributes)
```
Should raise in the fallback `find` case? It'd be surprising to ever get a `nil` result (for whatever reason, including a missing unique index) from these methods.

dhh replied:
Yeah, good point. Let's change that to find_by!.

Commit:
Raise an exception if no record can be found

Better than nil.

rafaelfranca commented on activerecord/test/cases/relations_test.rb:
```diff
@@ -1349,6 +1349,23 @@ def test_find_or_create_by!
+      Subscriber.create_or_find_by(nick: "bob", name: "the cat")
+    end
+  end
+
```
I think this is missing a test with `create_with` (to change the behavior of the `create`) and with `where` (to change the behavior of `find!`

I'm not sure if those method should be used with `create_or_find_by` but right now they can and I have no idea of the effect they would have, so better to test them to make sure it is what we expect.

dhh replied:
I don't see how that differs? `create_or_find_by` uses the same flow as `find_or_create_by`, except instead of an `||` we're using a `rescue`. But the flow is the same.

dhh replied:
I mean, it's not that we can't test it, just that I don't think it teaches us anything interesting. The test would be identical to the one for `test_find_or_create_by_with_create_with`.

rafaelfranca replied:
If it is the same behavior we want so I guess it is fine being tested in that test

eugeneius commented on activerecord/lib/active_record/relation.rb:
```diff
@@ -170,6 +159,41 @@ def find_or_create_by!(attributes, &block)
+    #   matching record, which will then return nil, rather than a record will the given attributes.
+    # * It relies on exception handling to handle control flow, which may be marginally slower. And
+    #
+    # This method will always returns a record if all given attributes are covered by unique constraints,
```
This isn't _totally_ true: another client could update or delete the row between the rejected INSERT and the subsequent SELECT. This race condition is complementary to the one in `#find_or_create_by`.

That caveat probably belongs in the "drawbacks" section, but even still I don't think we can make this claim here.

dhh replied:
Good point. Will add that point. It's a much rarer race condition in many apps, I'd say.

matthewd commented on activerecord/lib/active_record/relation.rb:
```diff
@@ -170,6 +159,42 @@ def find_or_create_by!(attributes, &block)
+    # #create returns in such situation.
+    def create_or_find_by(attributes, &block)
+      create(attributes, &block)
+    rescue ActiveRecord::RecordNotUnique
```
This needs `transaction(requires_new: true) do` around the `create` to work in an ongoing surrounding transaction (on at least PostgreSQL)

dhh replied:
Why is this not necessary for find_or_create_by?

matthewd replied:
Because it doesn't cause an SQL error and then attempt to recover. PostgreSQL remembers when an error has occurred inside a transaction, and disallows all further operations until that transaction has been rolled back.

sgrif replied:
If we are willing to require PG 9.5 or later for this (when using the PG adapter), we could just stick `ON CONFLICT DO NOTHING` on the end instead. This would let us wrap the whole thing in a transaction instead of just the create, which eliminates any possibility of race conditions.

matthewd commented on activerecord/lib/active_record/relation.rb:
```diff
@@ -170,6 +159,42 @@ def find_or_create_by!(attributes, &block)
+    def create_or_find_by!(attributes, &block)
+      create!(attributes, &block)
+    rescue ActiveRecord::RecordNotUnique
+      find_by!(attributes)
```
Do we want to `retry` here instead of raising if the find fails? Seems confusing for `create_or_find_by!` to raise RecordNotFound, just because of a racing delete.

dhh replied:
You can't retry this failure as it should only occur when you use attributes where not all of them are covered by unique constraints. It's more like a INVALID QUERY kind of failure.

matthewd replied:
Oh right, I was assuming the delete race, rather than operator error. I guess we could retry once, so we're safe from a poorly-timed simple delete. That way we'd only be tripped up by a racing `delete; insert; delete`.

dhh replied:
True. I'm not sure whether having just a single retry is a good warranty, though. Since it's hard for us to tell the difference between legit race condition and invalid query.

Commit:
Address the race condition between INSERT and SELECT

dhh commented:
TIL! Do you have an idea for a good test for that?

On Tue, Feb 13, 2018 at 4:49 PM, Matthew Draper <notifications@github.com>
wrote:

> *@matthewd* commented on this pull request.
> ------------------------------
>
> In activerecord/lib/active_record/relation.rb
> <https://github.com/rails/rails/pull/31989#discussion_r168049957>:
>
> > +    #
> +    # There are several drawbacks to #create_or_find_by, though:
> +    #
> +    # * The underlying table must have the relevant columns defined with unique constraints.
> +    # * A unique constraint violation may be triggered by only one, or at least less than all,
> +    #   of the given attributes. This means that the subsequent #find_by may fail to find a
> +    #   matching record, which will then raise an `ActiveRecord::NotFound` exception,
> +    #   rather than a record will the given attributes.
> +    # * It relies on exception handling to handle control flow, which may be marginally slower. And
> +    #
> +    # This method will always returns a record if all given attributes are covered by unique constraints,
> +    # but if creation was attempted and failed due to validation errors it won't be persisted, you get what
> +    # #create returns in such situation.
> +    def create_or_find_by(attributes, &block)
> +      create(attributes, &block)
> +    rescue ActiveRecord::RecordNotUnique
>
> Because it doesn't cause an SQL error and then attempt to recover.
> PostgreSQL remembers when an error has occurred inside a transaction, and
> disallows all further operations until that transaction has been rolled
> back.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/rails/rails/pull/31989#discussion_r168049957>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAAKtbTMH-JrUyizELV10shNLxOgesnAks5tUi2lgaJpZM4SElV8>
> .
>

tjschuck commented on activerecord/lib/active_record/relation.rb:
```diff
@@ -170,6 +159,47 @@ def find_or_create_by!(attributes, &block)
+    # * The underlying table must have the relevant columns defined with unique constraints.
+    # * A unique constraint violation may be triggered by only one, or at least less than all, 
+    #   of the given attributes. This means that the subsequent #find_by may fail to find a
+    #   matching record, which will then raise an `ActiveRecord::NotFound` exception,
```
`` `ActiveRecord::NotFound` `` needs `<tt>` for Rdoc code formatting, not backticks.

tjschuck commented on activerecord/lib/active_record/relation.rb:
```diff
@@ -170,6 +159,47 @@ def find_or_create_by!(attributes, &block)
+    # * A unique constraint violation may be triggered by only one, or at least less than all, 
+    #   of the given attributes. This means that the subsequent #find_by may fail to find a
+    #   matching record, which will then raise an `ActiveRecord::NotFound` exception,
+    #   rather than a record will the given attributes.
```
`will` => `with`

tjschuck commented on activerecord/lib/active_record/relation.rb:
```diff
@@ -170,6 +159,47 @@ def find_or_create_by!(attributes, &block)
+    # Attempts to create a record with the given attributes in a table that has a unique constraint
+    # on one or several of its columns. If a row already exists with one or several of these
+    # unique constraints, the exception such an insertion would normally raise is caught, 
+    # and the existing record with those attributes is sought found using #find_by.
```
`is sought found using #find_by` — "sought" should be deleted.

tjschuck commented on activerecord/lib/active_record/relation.rb:
```diff
@@ -170,6 +159,47 @@ def find_or_create_by!(attributes, &block)
+    # and the existing record with those attributes is sought found using #find_by.
+    #
+    # This is similar to #find_or_create_by, but avoids the problem of stale reads between the SELECT 
+    # and the INSERT, as that methods needs to first query the table, then attempt to insert a row 
```
`methods` => `method`

tjschuck reviewed (commented):
Documentation edits.

Commit:
Require new transaction in case unique constraint causes rollback

Otherwise PG will complain with "PG::InFailedSqlTransaction: ERROR:  current transaction is aborted, commands ignored until end of transaction block".

Thanks @matthewd

Commit:
Documentation fixes

Thanks @tjschuck

Commit:
Add changelog entry

Commit:
Remove trailing whitespace

AnalyzePlatypus commented:
Awesome!

subhashsaran commented:
Good stuff.

dhh commented:
Happy to explore that as a PG 9.5 level-up somehow. But the whole feature shouldn’t depend on it imo.

> On Feb 21, 2018, at 12:47, Sean Griffin <notifications@github.com> wrote:
> 
> @sgrif commented on this pull request.
> 
> In activerecord/lib/active_record/relation.rb:
> 
> > +    #
> +    # There are several drawbacks to #create_or_find_by, though:
> +    #
> +    # * The underlying table must have the relevant columns defined with unique constraints.
> +    # * A unique constraint violation may be triggered by only one, or at least less than all, 
> +    #   of the given attributes. This means that the subsequent #find_by may fail to find a
> +    #   matching record, which will then raise an `ActiveRecord::NotFound` exception,
> +    #   rather than a record will the given attributes.
> +    # * It relies on exception handling to handle control flow, which may be marginally slower. And
> +    #
> +    # This method will always returns a record if all given attributes are covered by unique constraints,
> +    # but if creation was attempted and failed due to validation errors it won't be persisted, you get what
> +    # #create returns in such situation.
> +    def create_or_find_by(attributes, &block)
> +      create(attributes, &block)
> +    rescue ActiveRecord::RecordNotUnique
> If we are willing to require PG 9.5 or later for this (when using the PG adapter), we could just stick ON CONFLICT DO NOTHING on the end instead. This would let us wrap the whole thing in a transaction instead of just the create, which eliminates any possibility of race conditions.
> 
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

sikachu commented:
(Hmm, @sgrif's comment was gone?)

@sgrif I'm 👍 on expanding this feature for PG 9.5+ in PG adapter. Other RDBMS can fallback to unique-key-constaint-exception-then-find.

yhirano55 commented:
I want to use these awesome class methods on Rails 5.2, so much.
Do you have any plans to backport to 5-2-stable?

bogdanvlviv commented:
I think we don't,  "New features are only added to the master branch and will not be made available in point releases." by [Maintenance Policy for Ruby on Rails](http://edgeguides.rubyonrails.org/maintenance_policy.html).

djezzzl commented:
Hi everybody,

As we start relying on the database constraints more, I'd suggest ([and we people like the idea](https://github.com/toptal/database_validations/issues/18)) to have the improved versions of `validates_uniqueness_of` and `belongs_to` from [database_validations](https://github.com/toptal/database_validations) gem. 

I just opened the issue https://github.com/rails/rails/issues/34650 where we can discuss the topic. 

*Any feedback is very appreciated!*

hunterae commented on activerecord/lib/active_record/relation.rb:
```diff
@@ -170,6 +159,47 @@ def find_or_create_by!(attributes, &block)
+    # and failed due to validation errors it won't be persisted, you get what #create returns in
+    # such situation.
+    def create_or_find_by(attributes, &block)
+      transaction(requires_new: true) { create(attributes, &block) }
```
I realize this wouldn't be create first, but is any reason not to make this implementation:

```
def create_or_find_by(attributes, &block)
    transaction(requires_new: true) { find_by(attributes) || create(attributes, &block) }
  rescue ActiveRecord::RecordNotUnique
    find_by!(attributes)
  end
```

Then the performance would be essentially identical to find_or_create_by, and it wouldn't have the overhead of catching the database exception.
